Merge with upstream changes

Thomas-Neill opened the pull request:
This should fix the segfault bugs we've been having.

We should merge https://github.com/HalloAppInc/gpb/pull/5 first.

ejabberd works when we use the merged versions of this and gpb

Commit:
Upgrade to rebar3

the-prksh committed:
Defined hooks for port compiler

the-prksh committed:
Change Plugin specs.

the-prksh committed:
Removes port_compiler

the-prksh committed:
Adds empty directory

the-prksh committed:
Adds empty directory

Commit:
Merge pull request #15 from prax-dev/master

Rebar3 Migration

Commit:
Fix tests error and compiling warnings.

Commit:
setup CI

Commit:
Update README.md

Commit:
improve CI (#20)

improve CI

Commit:
Update README.md

Commit:
remove unused proto files and fix benchmark (#21)

Commit:
fix rebar link issue (#22)

Commit:
fix ci (#24)

LiDoMin committed:
uint64 decode fail #19 (#23)

Signed-off-by: Dominic <389338435@qq.com>

Commit:
use enif_make_int64 to make signed 64 term (#25)

Commit:
add macos ci (#26)

add macos ci

Commit:
Update README.md

Commit:
fix issue [#11]

madninja committed:
move gpb to test only dependency (#28)

Commit:
align tests with gpb

dbegunkov committed:
support arm architect for m1 processor (#33)

Co-authored-by: Dmitry Begunkov <dbe@exante.eu>

Commit:
Merge remote-tracking branch 'upstream/master' into pull-changes

loukiano committed:
Fix random load_cache segfaults (#35)

vipin211 commented:
Can you please rebase changes made specifically by you into the last commit and describe your changes in the commit description?

Thomas-Neill commented:
> Can you please rebase changes made specifically by you into the last commit and describe your changes in the commit description?

Will do, didn't think about how this would be hard to review.

vipin211 commented:
> > Can you please rebase changes made specifically by you into the last commit and describe your changes in the commit description?
> 
> Will do, didn't think about how this would be hard to review.

Not hard to review, but for record keeping.

Commit:
repair after merging

Changed makefiles to use our settings, repair some mistakes made while merging, change some changes we made to follow new changes to the codebase, remove some new tests that were incompatible with functionality we changed

Thomas-Neill commented:
@vipin211 i've squashed changes on both repos now

vipin211 commented:
will ping you tomorrow to review in person. sorry for the delay.

kennedyjosh commented on .github/workflows/ci.yml:
```diff
@@ -7,6 +7,8 @@ on:
   pull_request:
     branches:
       - master
+  schedule:
+    - cron: '0 10 * * *' # Once per day at 10am UTC
```
remove this

kennedyjosh commented on .github/workflows/ci.yml:
```diff
@@ -15,7 +17,7 @@ jobs:
     strategy:
       matrix:
-        otp: ["23"]
+        otp: ["22", "23", "24"]
```
only need 23 and 24

kennedyjosh commented on .github/workflows/ci.yml:
```diff
@@ -35,3 +37,35 @@ jobs:
           make tests
+
+  macos:
+    name: Test on MacOS
```
we don't need to test this on mac

kennedyjosh commented on rebar.config:
```diff
@@ -10,6 +10,12 @@
     {"", compile, "escript enc compile"}
 ]}.
 
+{post_hooks, [
```
looks like this is here twice

Thomas-Neill replied:
this was removed in the commit after i merged

kennedyjosh reviewed (approved):
Same as gpb pr, we should watch this closely on s-test for at least a day

Commit:
fix ci.yml

Commit:
fix lock

loukiano commented on .gitignore:
```diff
@@ -15,5 +15,5 @@ cmake-build-debug
 log
-
-
+rebar.lock
```
I'm not sure we should ignore `rebar.lock` given that it overrides `rebar.config` unless specifically upgraded, I could see this causing some unintentional problems

kennedyjosh replied:
rebar specifically recommends not to ignore the rebar.lock file, good catch

balusu-murali commented:
> Same as gpb pr, we should watch this closely on s-test for at least a day

these kind of changes are hard to judge even in a single day.

loukiano commented on c_src/enif_protobuf.h:
```diff
@@ -41,31 +41,32 @@
 #if IS_DEBUG
 #define debug_out(fmt) do { \
-    fprintf(stderr, "%s:%d:%s(): " fmt "\r\n", __FILE__, __LINE__, __func__); \
+    printf("%s:%d:%s(): " fmt "\r\n", __FILE__, __LINE__, __func__); \
```
I kind of like these debuggers printing to stderr given that it's more reliably flushed. Personal preference but I would prefer keeping these as `fprintf(stderr...`

loukiano commented on c_src/enif_protobuf.h:
```diff
@@ -41,31 +41,32 @@
 #if IS_DEBUG
-#define return_exception(env, term) do {    \
+#define raise_exception(env, term) do { \
         printf("%s:%d:%s()\r\n", __FILE__, __LINE__, __func__); \
```
Error/Exception functions should definitely be printing to stderr imo

loukiano commented on c_src/enif_protobuf.h:
```diff
@@ -41,31 +41,32 @@
 
-#define return_error(env, term) do {    \
+#define return_error(env, term) do { \
         printf("%s:%d:%s()\r\n", __FILE__, __LINE__, __func__); \
```
super small nit but do we really need `\r` in these? I'm pretty sure `\n` does both newline and carriage return in C

loukiano replied:
Thought more about this and I think it's better this way; forgot that escape sequences are interpreted by the output processor, not C.

Thomas-Neill commented on c_src/ep_decoder.c:
```diff
@@ -896,53 +890,36 @@ fill_default(ErlNifEnv *env, ep_spot_t *spot)
 
-        } else if (field->type == field_enum) {
-            ep_enum_field_t    *efield;
-            efield = field->sub_node->v_fields;
```
@vipin211 
investigated and for enums, `v_fields` is a copy of `fields` which is sorted by integer value assigned to the enum. `fields` is sorted arbitrarily by the IDs of the atoms used for the fields. as a result i think our behavior is correct here (default should be the min. enum value) and theirs isn't. i'll take a look at the spec and make a PR upstream.

balusu-murali replied:
i made some fixes on our local copy to set the enum to the default field, to set bool default value as false etc.
https://github.com/HalloAppInc/enif_protobuf/commit/8848508735de20ea3287f436b5b70595ef307c85

why are we removing this code? dont think we should.
i am not sure about existing tests in this repo - but we should keep this code.

Thomas-Neill replied:
i think we should be going into the switch statement a few lines above to fill out default values now that that's been added; i forget what the proto compiler sets the occurrence options to by default, but if it's not already `repeated` or `defaulty` i can just remove the conditions for the above case in the if statement so that we get the same behavior.

Thomas-Neill replied:
checked again and everything's marked as `optional` - i'll make the change to just fill everything with the default value

vipin211 reviewed (approved):
I think we should try this code this week once we have various approvals.

Thomas-Neill committed:
fixed issue and added tests for it (#36)

Co-authored-by: Thomas Neill <thomas@Thomass-MacBook-Pro.local>

Commit:
remove rebar.lock from .gitignore

Commit:
Merge branch 'master' of github.com:jg513/enif_protobuf into pull-changes

Commit:
returned rebar.lock

balusu-murali commented on c_src/ep_encoder.c:
```diff
@@ -515,9 +515,11 @@ pack_bin_as_int64(ErlNifEnv *env, ERL_NIF_TERM term, ep_enc_t *enc)
-    if (enc->omit && val == 0) {
-        return RET_OK;
-    }
+
+    if (val == 0 && field->proto_v == 3 && field->o_type == occurrence_defaulty) {
+         return RET_OK;
+     }
+
```
why change this?

Thomas-Neill replied:
`enc->omit` was removed with the proto3 changes upstream; this is what they've changed this idiom to elsewhere in the file and this behavior seems correct to me.

balusu-murali replied:
did you get a chance to check if it works fine with the protobufs defined in our repo.
you can test some protobuf definitions in this repo standalone.. were you able to test our protobufs?

Thomas-Neill replied:
ok looked into it and since `proto_v` is set to `2` by default and `enc->omit` was also only set if `proto_v` was `3`, i don't think this case wasn't going off before, so this shouldn't change anything. i'm still going to test this case to make sure it works the same as on master though.

Thomas-Neill replied:
```
-record(m1, {a}).

encode_msg(Msg, Defs) ->
    encode_msg(Msg, Defs, [{string_as_list, true}, {with_utf8, true}]).

encode_msg(Msg, Defs, Opts) ->
    enif_protobuf:set_opts(Opts),
    enif_protobuf:load_cache(Defs),
    enif_protobuf:encode(Msg).


encodes_binary_correctly_test() ->
    <<8, 0>> = encode_msg( #m1{a = <<"0">>},
        [{{msg, m1}, [#?gpb_field{name = a, fnum = 1, rnum = #m1.a,
                type = int64, occurrence = optional,
                opts = [{ebin, true}]}]}]
        ).
```
This snippet passes on both `master` and this branch so I'm pretty confident this change isn't breaking anything.

balusu-murali replied:
i would use our own protobuf definitions - are you sure the proto_v is being picked up properly from our protobuf files?
the default occurrence type is optional for proto3.

balusu-murali commented on c_src/ep_encoder.c:
```diff
@@ -736,9 +738,9 @@ pack_element_packed(ErlNifEnv *env, ERL_NIF_TERM term, ep_enc_t *enc, ep_field_t
         if (field->ebin == TRUE) {
-            check_ret(ret, pack_bin_as_int64(env, term, enc));
+            check_ret(ret, pack_bin_as_int64(env, term, enc, field));
         } else {
-            check_ret(ret, pack_int64(env, term, enc));
+            check_ret(ret, pack_int64(env, term, enc, field));
```
did something change in these functions?

Thomas-Neill replied:
yes, i think they've changed the logic for when to not encode zero fields

balusu-murali commented on c_src/ep_decoder.c:
```diff
@@ -866,7 +866,11 @@ fill_default(ErlNifEnv *env, ep_spot_t *spot)
 
             case field_enum: {
-                *t++ = ((ep_enum_field_t *) field->sub_node->fields)->name;
+                ep_enum_field_t* efield = field->sub_node->v_fields;
+                int32_t target = 0;
+                ep_enum_field_t* zfield=bsearch(&target, efield, field->sub_node->v_size, sizeof(ep_enum_field_t), get_enum_compare_value);
+                if(zfield) *t++ = zfield->name;
+                else *t++ = ((ep_enum_field_t*)field->sub_node->fields)->name;
```
i thought the previous code was also doing something similar.. what are you trying to do here?

Thomas-Neill replied:
yes, this also should get the enum with a value of zero. the main difference is that this works if there's ever an enum assigned a negative value as you can't just grab the first one in that case.

balusu-murali replied:
we should never compile if there was a negative value in the enum in the first place.
dont follow why we have to do a binary search here in the first place.

Thomas-Neill replied:
it's technically legal but not recommended in the proto3 spec to have a negative value. i don't think we will ever be making a ton of enums with negative values, but when i made the pr upstream it's good to support that too.

balusu-murali replied:
ah - you are right about negative values. let us please do this in a separate diff.
let us keep this diff to just pulling in latest changes from master.

Thomas-Neill replied:
it has been merged on master upstream now

Commit:
change to always use default value, fix tests

balusu-murali commented on rebar.config:
```diff
@@ -1,5 +1,5 @@
 {deps, [
-    {gpb, ".*", {git, "https://github.com/HalloAppInc/gpb.git", {tag, "h1.2"}}}
+            {gpb, ".*", {git, "https://github.com/HalloAppInc/gpb.git", {branch, "merge-upstream"}}}
```
dont think these are linked. get rid of gpb changes here - i would prefer we upgrade gpb completely separately.

Thomas-Neill commented:
closing - the other PR is cleaner
